File: matching-engine/src/main/java/com/titanx/event/OrderEvent.java
package com.titanx.event;

import com.titanx.model.Order;
import com.titanx.model.Trade;

import java.time.Instant;

/**
 * ============================================================
 * TitanX OrderEvent — Phase 3: Event Sourcing
 * ============================================================
 *
 * Every state change in the system produces an immutable OrderEvent.
 * Events are appended to the `order_events` table in PostgreSQL.
 *
 * Why event sourcing?
 *  "Instead of updating a row in an `orders` table, we append events.
 *   Current state is a projection over events. This gives us:
 *   1. Crash recovery — replay events to reconstruct book state
 *   2. Full audit trail — required by MiFID II and RegNMS
 *   3. Time-travel queries — 'what was the order book at 14:32:07?'
 *   Bloomberg Terminal and ICE both use this pattern."
 *
 * Event types mirror the FIX protocol execution report reasons.
 */
public record OrderEvent(
        String eventId,
        EventType type,
        String orderId,
        String accountId,
        String symbol,
        Order  orderSnapshot,   // full order state at this point
        Trade  trade,           // non-null for FILLED events
        String rejectionReason, // non-null for REJECTED events
        long   latencyMicros,
        Instant occurredAt
) {
    public static OrderEvent received(Order order) {
        return build(EventType.ORDER_RECEIVED, order, null, null, 0);
    }

    public static OrderEvent resting(Order order) {
        return build(EventType.ORDER_RESTING, order, null, null, 0);
    }

    public static OrderEvent filled(Order order, Trade trade) {
        return build(EventType.ORDER_FILLED, order, trade, null, 0);
    }

    public static OrderEvent partiallyFilled(Order order, Trade trade) {
        return build(EventType.ORDER_PARTIALLY_FILLED, order, trade, null, 0);
    }

    public static OrderEvent cancelled(Order order) {
        return build(EventType.ORDER_CANCELLED, order, null, null, 0);
    }

    public static OrderEvent rejected(Order order, String reason) {
        return build(EventType.ORDER_REJECTED, order, null, reason, 0);
    }

    private static OrderEvent build(EventType type, Order order, Trade trade,
                                    String reason, long latencyMicros) {
        return new OrderEvent(
                java.util.UUID.randomUUID().toString(),
                type,
                order.orderId(),
                order.accountId(),
                order.symbol(),
                order,
                trade,
                reason,
                latencyMicros,
                Instant.now()
        );
    }
}

File: backend/src/event/OrderEvent.cpp
#include "titanx/event/OrderEvent.hpp"

namespace titanx {

OrderEvent OrderEvent::received(const Order& order) {
    return build(EventType::ORDER_RECEIVED, order, std::nullopt, "", 0);
}

OrderEvent OrderEvent::resting(const Order& order) {
    return build(EventType::ORDER_RESTING, order, std::nullopt, "", 0);
}

OrderEvent OrderEvent::filled(const Order& order, const Trade& trade) {
    return build(EventType::ORDER_FILLED, order, trade, "", 0);
}

OrderEvent OrderEvent::partiallyFilled(const Order& order, const Trade& trade) {
    return build(EventType::ORDER_PARTIALLY_FILLED, order, trade, "", 0);
}

OrderEvent OrderEvent::cancelled(const Order& order) {
    return build(EventType::ORDER_CANCELLED, order, std::nullopt, "", 0);
}

OrderEvent OrderEvent::rejected(const Order& order, const std::string& reason) {
    return build(EventType::ORDER_REJECTED, order, std::nullopt, reason, 0);
}

OrderEvent OrderEvent::build(EventType type, const Order& order,
                             const std::optional<Trade>& trade,
                             const std::string& reason,
                             long latencyMicros) {
    return OrderEvent{
        generate_uuid(),
        type,
        order.orderId,
        order.accountId,
        order.symbol,
        order,
        trade,
        reason,
        latencyMicros,
        SystemClock::now()
    };
}

nlohmann::json OrderEvent::to_json() const {
    nlohmann::json j;
    j["eventId"]         = eventId;
    j["type"]            = titanx::to_string(type);
    j["orderId"]         = orderId;
    j["accountId"]       = accountId;
    j["symbol"]          = symbol;
    j["orderSnapshot"]   = orderSnapshot.to_json();
    j["trade"]           = trade ? trade->to_json() : nullptr;
    j["rejectionReason"] = rejectionReason.empty() ? nullptr
                           : nlohmann::json(rejectionReason);
    j["latencyMicros"]   = latencyMicros;

    auto ms = std::chrono::duration_cast<std::chrono::milliseconds>(
        occurredAt.time_since_epoch()).count();
    j["occurredAt"] = ms;
    return j;
}

} // namespace titanx
